Sources/X509/TinyArray.swift

```diff
@@ -0,0 +1,323 @@
+        self.storage.append(newElement)
+    }
+    
+    @inlinable
+    mutating func append(contentsOf newElements: some Sequence<Element>){
+        self.storage.append(contentsOf: newElements)
+    }
+    
+    @discardableResult
+    @inlinable
+    mutating func remove(at index: Int) -> Element {
+        self.storage.remove(at: index)
+    }
+    
+    @inlinable
+    mutating func removeAll(where shouldBeRemoved: (Element) throws -> Bool) rethrows {
+        try self.storage.removeAll(where: shouldBeRemoved)
+    }
+    
+    @inlinable
+    mutating func sort(by areInIncreasingOrder: (Element, Element) throws -> Bool) rethrows {
+        try self.storage.sort(by: areInIncreasingOrder)
+    }
+}
+
+// MARK: - TinyArray.Storage "private" implementation
+
+extension TinyArray.Storage: Equatable where Element: Equatable {
+    @inlinable
+    static func ==(lhs: Self, rhs: Self) -> Bool {
+        switch (lhs, rhs) {
+        case (.one(let lhs), .one(let rhs)):
+            return lhs == rhs
+        case (.arbitrary(let lhs), .arbitrary(let rhs)):
+            // we don't use lhs.elementsEqual(rhs) so we can hit the fast path from Array
+            // if both arrays share the same underlying storage: https://github.com/apple/swift/blob/b42019005988b2d13398025883e285a81d323efa/stdlib/public/core/Array.swift#L1775
+            return lhs == rhs
+            
+        case (.one(let element), .arbitrary(let array)),
+            (.arbitrary(let array), .one(let element)):
+            guard array.count == 1 else {
+                return false
+            }
+            return element == array[0]
+            
+        }
+    }
+}
+extension TinyArray.Storage: Hashable where Element: Hashable {
+    @inlinable
+    func hash(into hasher: inout Hasher) {
+        // same strategy as Array: https://github.com/apple/swift/blob/b42019005988b2d13398025883e285a81d323efa/stdlib/public/core/Array.swift#L1801
+        hasher.combine(count)
+        for element in self {
+            hasher.combine(element)
+        }
+    }
+}
+extension TinyArray.Storage: Sendable where Element: Sendable {}
+
+extension TinyArray.Storage: RandomAccessCollection {
+    @inlinable
+    subscript(position: Int) -> Element {
+        get {
+            switch self {
+            case .one(let element):
+                guard position == 0 else {
+                    fatalError("index \(position) out of bounds")
+                }
+                return element
+            case .arbitrary(let elements):
+                return elements[position]
+            }
+        }
+    }
+    
+    @inlinable
+    var startIndex: Int {
+        0
+    }
+    
+    @inlinable
+    var endIndex: Int {
+        switch self {
+        case .one: return 1
+        case .arbitrary(let elements): return elements.endIndex
+        }
+    }
+}
+ 
+extension TinyArray.Storage {
+    @inlinable
+    init(_ elements: some Sequence<Element>) {
```

Comment:
This misses an optimization for the case where there is only one element.

Reply:
`self = .arbitrary([])` doesn't allocate and `self.append(contentsOf:)` has a fast path for a single element.